ktor-client/ktor-client-core/common/src/io/ktor/client/features/DefaultRequest.kt

```diff
@@ -37,3 +37,49 @@ public fun HttpClientConfig<*>.defaultRequest(block: HttpRequestBuilder.() -> Un
         block()
     }
 }
+
+/**
+ * Set the [baseUrl] to use it in sub-requests.
+ * To override the existing [baseURL] in a sub-request, the full [URL] is needed in the [HttpRequestBuilder].
+ *
+ * The given [baseUrl] cannot have a query or a fragment part.
+ * The given [baseUrl] cannot be the default [URLBuilder].
+ */
+public fun HttpRequestBuilder.baseURL(baseUrl: String) {
+    baseURL(URLBuilder().takeFrom(baseUrl))
+}
+
+/**
+ * Set the [baseUrlBuilder] to use it in sub-requests.
+ * To override the existing [baseURL] in a sub-request, the full [URL] is needed in the [HttpRequestBuilder].
+ *
+ * The given [baseUrlBuilder] cannot have a query or a fragment part.
+ * The given [baseUrlBuilder] cannot be the default [URLBuilder].
+ */
+public fun HttpRequestBuilder.baseURL(baseUrlBuilder: URLBuilder) {
+    require(baseUrlBuilder.parameters.build() == Parameters.Empty && baseUrlBuilder.fragment.isEmpty()) {
+        "The baseURL cannot have a query or a fragment"
+    }
+
+    url {
+        val defaultURLBuilder = URLBuilder()
+        val isDefault = this == defaultURLBuilder
+        val onlyPathSet = startsWith(defaultURLBuilder)
+        if(!isDefault && onlyPathSet) {
```

Comment:
Why do we do nothing if `!onlyPathSet`?

Reply:
Keep in mind, the default request feature is called after! the user customized the url.
If the user uses `a.com/api` as baseURL, and requests `/some`, the `url` in `HttpRequestBuilder` is `http://localhost/some`, because of the default values in `URLBuilder`. So you have to check, if the `url` is a relative URL. If the url is relative (`onlyPathSet`), the baseURL is added. If `(!onlyPathSet`) the url is "complete", has a host, so the baseURL feature is not used.

Reply:
Maybe I should rename the variable to `isRelativeURL`